File: algorithms/Python/depthFirstSearch.py
class Node:
    def __init__(self, name):
        self.children = []
        self.name = name

    def addChild(self, name):
        self.children.append(Node(name))

    # O(v + e) Time | O(v) Space
    def depthFirstSearch(self, array):
        array.append(self.name) # append to our array the nodes name
        for child in self.children:
            child.depthFirstSearch(array)
        return array # useful only for first call of dfs on recursive stack

File: algorithms/Javascript/depthFirstSearch.js
class Node {
  constructor(name) {
    this.name = name;
    this.children = [];
  }

  addChild(name) {
    this.children.push(new Node(name));
    return this;
  }

  // O(v + e) Time | O(v) Space
  depthFirstSearch(array) { // takes empty result-array as argument
    array.push(this.name); // add current node node to array
    for (let child of this.children) { // loops through and assigns variable to each child from this.children array
      child.depthFirstSearch(array); // recursively call dfs on children, adding each to call stack
    } // reaches leaf of tree before bubbling up call stack
    return array; // returns the first dfs call
  }
}
